File: oo/src/test/java/oo/ep/M0Test.java
package oo.ep;

import org.junit.Assert;
import org.junit.Test;

public class M0Test {
    public interface TestTemplate  {
        default void test() {
            Exp expr1 = new Add(new Lit(1.0), new Lit(2.0));
            Assert.assertEquals(3.0, expr1.eval(), 0.0);

            Exp expr2 = new Lit(2.0);
            Assert.assertEquals(2.0, expr2.eval(), 0.0);

            Assert.assertEquals(3.0, new Add(new Lit(1.0), new Lit(2.0)).eval(), 0.0);
            Assert.assertEquals(5.0, new Lit(5.0).eval(), 0.0);
        }
    }

    private static class ActualTest implements TestTemplate {}

    @Test
    public void testTest() { new ActualTest().test(); }
}

File: coco_scala/src/test/scala/exp/m0/M0Test.scala
package exp.m0

import org.scalatest.FunSuite

trait TestTemplate[FT] extends Factory[FT] {
  val suite:FunSuite
  import suite._

  def test() : Unit = {
    val one = lit(3)
    val two = lit(8)
    val addOneTwo = add(one, two)
    assert(addOneTwo.eval === 11)
  }
}

class M0Test extends FunSuite { self =>

  object ActualTest extends TestTemplate[finalized.Exp] with finalized.Factory {
    val suite: FunSuite = self
  }

  test("M0") {
    ActualTest.test()
  }
}
